src/filters.cpp

```diff
@@ -868,16 +868,190 @@ InternalValue Slice::Batch(const InternalValue& baseVal, RenderContext& context)
-StringFormat::StringFormat(FilterParams, StringFormat::Mode)
+StringFormat::StringFormat(FilterParams params, StringFormat::Mode mode)
+: m_mode{mode}
 {
-
+    if(m_mode == Mode::PythonMode)
+    {
+        ParseParams({}, params);
+        m_params.kwParams = std::move(m_args.extraKwArgs);
+        m_params.posParams = std::move(m_args.extraPosArgs);
+    }
 }
 
-InternalValue StringFormat::Filter(const InternalValue&, RenderContext&)
+InternalValue StringFormat::Filter(const InternalValue& baseVal, RenderContext& context)
 {
+    if(m_mode == PythonMode)
+        return PythonFormat(baseVal, context);
+
     return InternalValue();
 }
 
+namespace {
+
+using FormatContext = fmt::format_context;
+using FormatArgument = fmt::basic_format_arg<FormatContext>;
+
+template<typename ResultDecorator>
+struct FormatArgumentConverter : visitors::BaseVisitor<FormatArgument>
+{
+    using result_t = FormatArgument;
+
+    using BaseVisitor::operator();
+
+    FormatArgumentConverter(const RenderContext* context, const ResultDecorator& decorator)
+        : m_context(context), m_decorator(decorator)
+    {}
+
+    result_t operator()(const ListAdapter& list) const
+    {
+        return make_result(Apply<PrettyPrinter>(list, m_context));
+    }
+
+    result_t operator()(const MapAdapter& map) const
+    {
+        return make_result(Apply<PrettyPrinter>(map, m_context));
+    }
+
+    result_t operator()(const std::string& str) const
+    {
+        return make_result(str);
+    }
+
+    result_t operator()(const nonstd::string_view& str) const
+    {
+        return make_result(std::string(str.data(), str.size()));
+    }
+
+    result_t operator()(const std::wstring& str) const
+    {
+        return make_result(ConvertString<std::string>(str));
+    }
+
+    result_t operator()(const nonstd::wstring_view& str) const
+    {
+        return make_result(ConvertString<std::string>(str));
+    }
+
+    result_t operator()(double val) const
+    {
+        return make_result(val);
+    }
+
+    result_t operator()(int64_t val) const
+    {
+        return make_result(val);
+    }
+
+    result_t operator()(bool val) const
+    {
+        return make_result(val ? "true"s : "false"s);
+    }
+
+    result_t operator()(EmptyValue) const
+    {
+        return make_result("none"s);
+    }
+
+    result_t operator()(const Callable&) const
+	{
+		return make_result("<callable>"s);
+	}
+
+    template<typename T>
+    result_t make_result(const T& t) const
+    {
+        return fmt::internal::make_arg<FormatContext>(m_decorator(t));
+    }
+
+
+    const RenderContext* m_context;
+    const ResultDecorator& m_decorator;
+};
+
+template<typename T>
+using NamedArgument = fmt::internal::named_arg<T, char>;
+
+using ValueHandle = nonstd::variant<
+    bool,
+    std::string,
+    int64_t,
+    double,
+    NamedArgument<bool>,
+    NamedArgument<std::string>,
+    NamedArgument<int64_t>,
+    NamedArgument<double>
+>;
+using ValuesBuffer = std::vector<ValueHandle>;
```

Comment:
What if you replace vector with deque here?

Reply:
Arghh! Github merged master into my branch, but it lost commit, which I wanted to try. I will try it again and, if it won't work, I will try deque.